packages/shared/src/hooks/usePrivilegedSession.ts

```diff
@@ -0,0 +1,98 @@
+import { useMemo } from 'react';
+import { useMutation, useQuery } from 'react-query';
+import { LoginFormParams } from '../components/auth/LoginForm';
+import { getNodeValue, ValidateResetPassword } from '../lib/auth';
+import {
+  AuthFlow,
+  InitializationData,
+  initializeKratosFlow,
+  submitKratosFlow,
+} from '../lib/kratos';
+import useLogin from './useLogin';
+import usePersistentContext from './usePersistentContext';
+import { useToastNotification } from './useToastNotification';
+
+interface ResetFormParams {
+  password: string;
+}
+
+interface UsePrivilegedSession {
+  showVerifySession?: boolean;
+  onCloseVerifySession: () => void;
+  settings: InitializationData;
+  onUpdatePassword: (params: ResetFormParams) => unknown;
+  onPasswordLogin?: (params: LoginFormParams) => void;
+  onSocialLogin?: (provider: string) => void;
+}
+
+interface UsePrivilegedSessionProps {
+  onResetSuccessful?: () => void;
+  onSessionVerified?: () => void;
+}
+
+export const VERIFY_SESSION_KEY = 'verify_session';
+
+const usePrivilegedSession = ({
+  onResetSuccessful,
+  onSessionVerified,
+}: UsePrivilegedSessionProps = {}): UsePrivilegedSession => {
+  const { displayToast } = useToastNotification();
+  const [verifySessionId, setVerifySessionId] =
+    usePersistentContext(VERIFY_SESSION_KEY);
+  const { data: settings } = useQuery(['settings'], () =>
+    initializeKratosFlow(AuthFlow.Settings),
+  );
+
+  const { mutateAsync: resetPassword } = useMutation(
+    (params: ValidateResetPassword) => submitKratosFlow(params),
+    {
+      onSuccess: async ({ redirect }) => {
+        if (redirect) {
+          const url = new URL(redirect);
+          const returnTo = new URL(url.searchParams.get('return_to'));
+          const flowId = returnTo.searchParams.get('flow');
+          return setVerifySessionId(flowId);
+        }
+
+        return onResetSuccessful?.();
+      },
+    },
+  );
```

Comment:
I'm not a 100% sure this should be in this file.
It's a specific action and can work independently from the privileged session

Reply:
Totally agree, just thought this will probably be used in some areas as well and will definitely use the privileged session. But we can take this out of here, for sure.